src/pushsource/_impl/backend/registry_source.py

```diff
@@ -0,0 +1,143 @@
+import logging
+
+from ..source import Source
+from ..model import (
+    ContainerImagePushItem,
+    SourceContainerImagePushItem,
+    ContainerImagePullInfo,
+    ContainerImageTagPullSpec,
+    ContainerImageDigestPullSpec,
+)
+from ..utils.containers import (
+    get_manifest,
+    inspect,
+    MT_S2_V2,
+    MT_S2_V1,
+    MT_S2_V1_SIGNED,
+    MT_S2_LIST,
+)
+
+from ..helpers import list_argument
+
+LOG = logging.getLogger("pushsource")
+
+
+class RegistrySource(Source):
+    """Uses URIs of container images as source for push items."""
+
+    def __init__(
+        self,
+        images_str,
+        dest_repos=None,
+        dest_signing_key=None,
+    ):
+        """Create a new source.
+
+        Parameters:
+            dest_repos str,
```

Comment:
I don't think this and the destination_tag stuff should be included, it still seems arbitrary and overly specialized for your current use-case (by the way, is prototype code using this available already? Can you share if so?)

If we compare all the backends now:

- errata: the underlying data source (ET) knows the destination, so it's filled in automatically.
- staged: the underlying data source (filesystem) knows the destination, so it's filled in automatically.
- koji: the underlying data source does not know the destination. There is one parameter, 'dest', which can be used to fill in the destination if all push items are going to the same place. The resulting 'dest' in the output is an arbitrary string, the source does not require it to be any particular format or represent any specific thing.  For any more complex cases, the caller has to set the destination after getting the push items.
- registry: the underlying data source does not know the destination. There are two parameters, 'dest_repos' and 'images_str', influencing the destination; it'll be set on all push items by combining data from these parameters in  such a way that dest_repos is applied to *all* push items and different tags from images_str will be applied to *each* push item.  The resulting 'dest' in the output is specifically a container image repo:tag.  For any more complex cases, the caller has to pass a fake destination, and can then rewrite the destination after getting the push items.

koji and registry shouldn't work so differently, and registry shouldn't assume that we're always gathering push items for push to another registry (which seems implied by the repo:tag output format).

I think "if in doubt, leave it out" applies here.  Can you please either make registry work exactly like koji, or even drop the destination handling entirely and require it to be handled by the caller?

Reply:
yes, for koji source tags are externally provided. Also that's why it's not possible to map exact source tags to exact destination tags. You can only set destination tags for all push items. I don't find it very useful, that's why I implemented it this way. 
Or do you have another idea how to do the mapping?